napari/layers/utils/layer_utils.py

```
@@ -684,6 +684,43 @@ def features_to_pandas_dataframe(features: Any) -> pd.DataFrame:
     return features
 
 
+def _features_from_layer(
+    *,
+    features: Optional[Union[Dict[str, np.ndarray], pd.DataFrame]] = None,
+    properties: Optional[Union[Dict[str, np.ndarray], pd.DataFrame]] = None,
+    property_choices: Optional[Dict[str, np.ndarray]] = None,
+    num_data: Optional[int] = None,
+) -> Tuple[pd.DataFrame, pd.DataFrame]:
+    """Coerces a layer's keyword arguments to feature values and defaults tables."""
+    if properties is not None or property_choices is not None:
+        features = _features_from_properties(
+            properties=properties,
+            property_choices=property_choices,
+            num_data=num_data,
+        )
+    else:
+        features = _validate_features(features, num_data=num_data)
+    defaults = pd.DataFrame(
+        {
+            name: _get_column_default_value(column)
+            for name, column in features.items()
+        },
+        index=range(1),
+    )
+    return features, defaults
+
+
+def _get_column_default_value(column: pd.Series) -> Optional:
+    """Get the default value from a feature column."""
+    if column.size > 0:
+        return column.iloc[-1]
+    if isinstance(column.dtype, pd.CategoricalDtype):
+        choices = column.dtype.categories
+        if choices.size > 0:
+            return choices[0]
```

Should this be nested in the `if` above? I'm not super familiar with pandas categorical type, but I would assume a categorical column that has values would still have size > 0.

---

I tried to mostly follow the existing logic from `get_current_properties`:https://github.com/napari/napari/blob/5b3389d73526b28554c33adf5ec857d1560a7f8f/napari/layers/utils/layer_utils.py#L326

where if there's any data (i.e. size > 0), then the default value is the last value (regardless if the type is categorical or not) and we're done. 

If there is no data, then the default for categoricals is the first choice (now the first category value in the dtype). The check for `choices.size > 0` is to guard against the weird, but possible, case when we have a categorical with no values and no choices (`s = pd.Series([], dtype=pd.CategoricalDtype([]))`). Maybe at that point, we should just raise an exception instead and inform the user they are in a bad place?

Otherwise, we can't define a default value and return `None`.